Sparrow dom/search

sparrowDom opened the pull request:
### Checklist:

- [x] Test your work and double check you didn't break anything
- [ ] Update any relevant READMEs and [docs](https://github.com/OriginProtocol/docs)
- [x] **Important!** clear `announcements-search.json` of filter configuration before merging

### Description:
- Creates a search bar just under the navigation bar that lets you search for different Listings. Only the query typed in the input bar is connected to GraphQl. Other filters (listingType, price, dates....) are not yet connected. 
- Listings are displayed on a search result page where multiple filters can be applied
- Search filters are configurable for each listing type. Search for `{listingType}-search.json`schema files
- Currently only `announcements` listing type has search filters configured (for testing purposes) see `announcements-search.json` configuration file
- Filters are not yet implemented on GraphQl backend so full circle can not be tested
- Search filter states are kept in redux and forwarded to `origin-js` whenever an `Apply` buttons is clicked in the filter dropdown. 
- When testing with dependant `origin-js` PR it is easiest to change [this line](https://github.com/OriginProtocol/origin-js/pull/421/files#diff-1fdf421c05c1140f6d71444ea2b27638R12) to `localhost` and run Appolo server locally with command: `origin-box origin-js$ docker exec -w /app/daemon/indexing origin-js node apollo/index.js`
- the filters passed to the `origin-js` can be observed in the chrome network inspector. See below screenshot 
<img width="1920" alt="screenshot 2018-09-07 00 15 26" src="https://user-images.githubusercontent.com/579910/45188085-4c6ba780-b233-11e8-828e-e630fc4a1317.png">


**Dependant PR** https://github.com/OriginProtocol/origin-js/pull/421
Connected to issue: https://github.com/OriginProtocol/origin-dapp/issues/313#issuecomment-411497217

franckc commented on src/components/listings-grid.js:
```diff
@@ -23,7 +23,8 @@ class ListingsGrid extends Component {
     const { listingsPerPage } = this.state
     const { contractFound, listingIds, hideList } = this.props
-    const pinnedListingIds = [0, 1, 2, 3, 4]
+    //const pinnedListingIds = [0, 1, 2, 3, 4]
```
Those hardcoded pinnedListingIds are used for "featuring" listings.
We'll need to replace this with a programmatic way (we discussed having some sort of admin UI allowing Origin employees to feature/unfeature listings).
But for now, we should probably keep the list ?

sparrowDom replied:
Yeah this is a mistake from my part. These ones shouldn't be commented out. Thanks for the catch

franckc commented on src/utils/listingSchema.js:
```diff
@@ -0,0 +1,54 @@
+const listingSchema = {
+	listingTypes: [
```
nit: fix indentation for this file. I think our convention is to use 2 spaces.

franckc commented on src/utils/listingSchema.js:
```diff
@@ -0,0 +1,54 @@
+	]
+}
+
+export default listingSchema
```
nit: add a newline at end of file.

franckc commented on translations/all-messages.json:
```diff
@@ -46,12 +46,6 @@
   "listing-card-prices.ethereumCurrencyAbbrev": "ETH",
   "listing-card-prices.sold": "Sold",
   "listing-card.featured": "Featured",
-  "listing-create.forSaleLabel": "For Sale",
```
Why did those strings get deleted ? Don't we still need them for the listing creation page ?

sparrowDom replied:
Good question will look into that. It could be that I have removed the hardcoded translations from `listing-create` into a separate file, because `searchbar` also requires that schema/translations. Just so that we do not have the data duplicated...

franckc commented on translations/messages/src/components/search/searchbar.json:
```diff
@@ -0,0 +1,18 @@
+    "id": "searchbar.delasNearMe",
+    "defaultMessage": "Deals Near Me"
+  }
+]
```
super nit: add newline

sparrowDom replied:
I think this file is automatically created by the translation plugin. Will add the new line anyway :)

franckc commented:
Adding @micahalcorn as a reviewer since he is the DApp expert.

@sparrowDom Let us know when this is ready for another round of review.

micahalcorn commented on src/css/app.css:
```diff
@@ -957,92 +1019,92 @@ a:hover {
 }
 
-.navbar .navbar-nav .nav-link img.add-listing {
+.navbar.navigation-bar .navbar-nav .nav-link img.add-listing {
```
Just a comment that Josh is not a fan of namespacing. I realize that there is already a lot of it in this file because I was once following a convention from the website repo and haven't made it a priority to refactor all of it. But we should favor cascading and limit namespacing as much as possible.

sparrowDom replied:
Thank @micahalcorn will convert it to cascading. I haven't done CSS in over 8 years so all these comments are super helpful.
I am happy to receive as many PR comments as possible. Since I want to get up to speed with the latest CSS best practices ASAP.

sparrowDom commented on src/components/layout.js:
```diff
@@ -2,11 +2,13 @@ import React, { Fragment } from 'react'
   <Fragment>
     <main className="d-flex flex-column">
       <NavBar />
+      <SearchBar />
```
@micahalcorn yes this is a work in progress. I have created a PR so I can more easily diff all the changes made in the branch with the develop branch. 

So this is totally a mistake on my part. Please ignore this PR and will ping you once it is ready for review

micahalcorn commented on src/components/search/counter-filter.js:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    this.handleOnClickAdd = this.handleOnClickAdd.bind(this)
+    this.handleOnClickSubstract = this.handleOnClickSubstract.bind(this)
```
While @franckc can inevitably produce some obscure, ancient reference to "substract" being a word, I recommend that all instances of this be replaced with "subtract". ➖

micahalcorn commented on src/components/search/filter-group.js:
```diff
@@ -0,0 +1,158 @@
+          onChildUnMounted={this.handleFilterUnMounted}
+        />
+      )
+    } else if (filter.type == 'price') {
```
Let's use strict [in]equality everywhere unless there is a compelling reason not to. 🔒

micahalcorn commented on src/components/search/search-result.js:
```diff
@@ -0,0 +1,219 @@
+    this.handleComponentUpdate(undefined)
+
+    // Keep dropdown opened when user clicks on any element in the dropdownw
+    $(document).on('click', '#search-filters-bar .dropdown-menu', e => {
```
Can this be accomplished using [a dropdown event listener](https://getbootstrap.com/docs/4.0/components/dropdowns/#events)? Maybe not. How about an `onClick` handler on the relevant React components? I would like to get away from jQuery at some point, but it doesn't have to start here.

sparrowDom replied:
Afaik returning false from `hide.bs.dropdown ` event should stop the dropdown from closing. But you would still need to use jQuery to attach to that event. I am not happy with this solution also, but couldn't find a better one.

micahalcorn commented on src/components/search/search-result.js:
```diff
@@ -0,0 +1,219 @@
+
+  render() {
+    return (
+      <div>
```
Can we use a `Fragment` here instead of an empty `div`?

sparrowDom replied:
Aaa thanks, wasn't aware `Fragments` are there to avoid using unnecessary divs.

micahalcorn commented on package.json:
```diff
@@ -21,8 +22,11 @@
     "react": "^16.3.2",
     "react-avatar-editor": "^11.0.2",
+    "react-dates": "17.2.0",
     "react-dom": "^16.3.2",
```
These dependencies all look good. I'll take another pass at the entire PR once the conflicts are resolved.

franckc commented on public/schemas/searchFilters/announcements-search.json:
```diff
@@ -0,0 +1,78 @@
+{
+  "type": "filterCreationSchema",
+  "listingType": "announcements",
```
The content of this file looks to me like filters for the housing listing type. But the filename and listingType refer to announcements ?

sparrowDom replied:
Yeah, it doesn't make sense currently.  I briefly mentioned this in the Pull Request's description. All the filter types are in the `announcements-search.json` just for testing purposes. Before I merge this PR I will clear the `announcements-search.json` file of the filters. UI and logic of the filters should be working, but since they are not connected to GraphQL yet they really shouldn't be displayed at all.

franckc commented on public/schemas/searchFilters/announcements-search.json:
```diff
@@ -0,0 +1,78 @@
+  "items": [
+    {
+      "title": {
+        "id": "schema.forSale.price",
```
It is a bit confusing reading this file and seeing those references to strings from the UI JSON schemas that are not of the same type as the filter. Since the filters apply to fields that are collected during listing creation, there should not be any reason to not use those ?
e.g. for the "housing" filter, use strings from the [housing UI schema](https://github.com/OriginProtocol/origin-dapp/blob/master/public/schemas/housing.json). for the "announcement" filters, use strings from the [announcement UI schema](https://github.com/OriginProtocol/origin-dapp/blob/master/public/schemas/announcements.json) , etc...

sparrowDom replied:
Yes that makes sense. We will need to be careful of that when creating the real filters and not just these ones that are intended for testing UI. 

Have changed the strings in the `announcements-search.json` so that we do not forget about that in the future.

franckc commented on src/components/listing-create.js:
```diff
@@ -36,66 +37,14 @@ class ListingCreate extends Component {
-        img: 'announcements.jpg'
-      }
-    ]
+    console.log(this.schemaList1)
```
remove those debugging statements

franckc commented on translations/messages/src/components/search/searchbar.json:
```diff
@@ -0,0 +1,18 @@
+    "defaultMessage": "New Listings"
+  },
+  {
+    "id": "searchbar.delasNearMe",
```
typo: "deals" not "delas"

sparrowDom replied:
nice catch thanks!

sparrowDom commented:
@franckc @micahalcorn all the conflicts, formatting errors and current comments are resolved/addressed. Please continue with review when you have time

micahalcorn commented:
@sparrowDom I made a handful of changes.

1) Code style 😇 
2) Styling tweaks
3) Repair `MultipleSelectionFilter` uncheck UX
4) Improved 💲 symbol placement and controlled the component - There may be a reason that you were using jQuery for this. I'm not testing with a backend, so I don't know if my change breaks it for some reason.

franckc commented:
Tested this DApp PR (while running origin-js on the sparrowDom/search branch) and got search to work.

A few questions/comments:
1. When I got on the default page (e.g. http://localhost:3000/#/), there was no listing. I think we should preserve the current behavior where we show list of listings with "pinned" listings first.
2. I noticed this warning in the JS console when search results get displayed:
```
warning.js:33 Warning: Each child in an array or iterator should have a unique "key" prop.
Check the render method of `SearchResult`. See https://fb.me/react-warning-keys for more information.
    in Connect(InjectIntl(FilterGroup)) (created by SearchResult)
    in SearchResult (created by Route)
    in Route (created by withRouter(SearchResult))
    in withRouter(SearchResult) (created by InjectIntl(withRouter(SearchResult)))
    in InjectIntl(withRouter(SearchResult)) (created by Connect(InjectIntl(withRouter(SearchResult))))
    in Connect(InjectIntl(withRouter(SearchResult))) (created by Route)
```
3. Until they are implemented, should we deactivate the links that are on right side of search bar (For Sale, New Listing, Deal near me) ?
4. In the price filter, since it is not specific to the Housing category but applies to any category, should we replace the $/night by just $ ?

sparrowDom commented:
Thanks @micahalcorn for the changes and the fixes, for catching the MultipleSelectionFilter error, making a non jQuery solution to the price filter and search-bar UI enhancement. Maybe only the style fixes part I would prefer to be done in a separate branch -> since it pulled in a lot of files from the codebase that this PR has not touched. 

@franckc:
1. thanks for pointing out the problem with default listings now showing up. It was a bug and is now fixed. 
2. I do not see the error with children requiring `key` property. It is a bug I have fixed Friday I think. Maybe the branch is not up to date? 
3. Also I think I have deactivated the "For Sale, New Listing, Deal near me" on Thursday or Friday so they should not show up anymore. 
4. Good point. Have added the ability to configure the PriceFilter's unit as `priceUnit` from the configuration schema (commit https://github.com/OriginProtocol/origin-dapp/pull/396/commits/11574bcddcc5f2ff8aac0bdfbfca6cef09a11661 )

sparrowDom commented:
@franckc @micahalcorn do you want to do another pass or do we merge it?

micahalcorn commented:
@sparrowDom @franckc do we have a good idea of which UI elements in this PR will need to be hidden for the v1 deployment? Or are we still waiting to see which features are ready to be supported on the backend?

franckc commented:
@sparrowDom Did another round of testing and looks good ! :)
As a separate PR, can you add the ENV variable for discovery server ?

@micahalcorn Domen hid all the elements that should be excluded from V1 already. Then the search filters are completely configurable (see https://github.com/OriginProtocol/origin-dapp/tree/sparrowDom/search/public/schemas/searchFilters) so for those will be trivial to only keep the ones that are finished by V1.

I think we are good for merging this PR. @micahalcorn ok by you ?

sparrowDom commented:
sounds good @franckc will add configuration in a separate PR.
@micahalcorn yeah let me just do one more commit ( https://github.com/OriginProtocol/origin-dapp/pull/396/commits/bf02215ef427df140d752dcccb1e8699b46949f2 ) that removes the UI filters from `announcements-search.json`. Now when each filter is implemented in the backend we will just input the appropriate configuration to json files like @franckc says.
